### PR TITLE
EQ IIR: Add support for response reconfigure during streaming

### PR DESCRIPTION
**Pull request by singalsu**
This patch adds capability to equalizer to switch response during
audio playback or capture. The glitches are minimized by check and
apply of the filter configuration before each copy() operation.

The existing only halfway implemented SOF_CTRL_CMD_ENUM for quick
stored presets switching during streaming is removed for now since
there is no driver support.

The internal code cleanup includes extracting two functions from
eq_iir_setup() for check and initialize of coefficients and
initialize of filter delay lines.

To minimize number of duplicated functions in generic and HiFi3 version
a new module iir_generic.c is created for the filter core. After the
change iir.c contains only the common code for both implementations.

Finally, some trace messages are cleaned up to easier to understand
text.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
I get checkpatch warnings from trace functions usage. What's the recommended practice. Should all traces be trace_x_with_ids() type instead of trace_x()? Also should the trace print multiple line split strings be in single line now?

**Comment by paulstelian97** [comment]
@singalsu Indeed with the recent Checkpatch changes it is recommended to not split long strings. However some splits are legitimate.

Please do not split individual phrases which one may grep for, like "request for non-defined filter" or "Illegal number of biquads". You may split and safely ignore the warning if you split at some ":", "." or format specifier though.

The one with "%s..." should be completely ignored, the current trace system doesn't support %s at all.

Do focus on the alignment with open parenthesis too.

**Comment by singalsu** [comment]
@paulstelian97 Thanks for help! I'll re-format this PR.

**Comment by jajanusz** [comment]
@paulstelian97 @singalsu Which splits are legitmiate?

**Comment by paulstelian97** [comment]
@jajanusz I'd argue all of the ones present now are good enough because I doubt someone will try to grep for the whole message in order to debug. But if you argue otherwise and there should be no splitting at all then fair enough.

@singalsu Is there anything that can be done about the line too long in eq_iir.c:883? Or would that break an alignment for no real gain?

**Comment by singalsu** [comment]
> @singalsu Is there anything that can be done about the line too long in eq_iir.c:883? Or would that break an alignment for no real gain?

a new pointer "cn" could store "cd->config_new" and the arguments could be "cn" and "cn->size". I've sometimes done such to get all green check marks. Now that we get warning for every trace I don't now if aiming for all green is feasible.

I tried to split size into third argument line but then checkpatch warns about xx->yy->zz pointers usage. It's strange that keeping it in single line doesn't give such warning.

What would you prefer?

**Comment by singalsu** [comment]
Also I don't like myself the longer than earlier trace macro name that triggered these many issues with long or split strings lines. Why could not the trace_eq() be converted to do the same as trace_eq_with_ids(). Why was the old no-ids trace preserved?

**Comment by jajanusz** [comment]
@singalsu @paulstelian97 I think it doesn't matter what we think in that regard. One of the project goals is to be compatible with kernel coding style and it clearly states: `However, never break user-visible strings such as printk messages, because that breaks the ability to grep for them.`

https://www.kernel.org/doc/html/latest/process/coding-style.html#breaking-long-lines-and-strings

**Comment by paulstelian97** [comment]
> @singalsu @paulstelian97 I think it doesn't matter what we think in that regard. One of the project goals is to be compatible with kernel coding style and it clearly states: `However, never break user-visible strings such as printk messages, because that breaks the ability to grep for them.`
> 
> https://www.kernel.org/doc/html/latest/process/coding-style.html#breaking-long-lines-and-strings

So we should only be allowed to split at format specifiers (because you can't grep for the value anyway). So only at %d or something. Got it.

**Comment by singalsu** [comment]
The push that I just did addresses the trace strings split and long lines recommendation. I shortened the local trace macro names to keep sufficient verbosity. Hope it's OK now.

File: src/audio/eq_iir/eq_iir.c [hunk 1]
```diff
@@ -31,30 +31,31 @@
 #include <stdint.h>
 
 /* tracing */
-#define trace_eq(__e, ...) trace_event(TRACE_CLASS_EQ_IIR, __e, ##__VA_ARGS__)
```
Comment (paulstelian97) [inline src/audio/eq_iir/eq_iir.c]: Why do we rename these? (also, if there's a good reason please split this rename into a separate commit)

File: src/audio/eq_iir/eq_iir.c [hunk 2]
```diff
@@ -31,30 +31,31 @@
...
 
-#define tracev_eq(__e, ...) tracev_event(TRACE_CLASS_EQ_IIR, __e, ##__VA_ARGS__)
-#define tracev_eq_with_ids(comp_ptr, __e, ...)			\
+#define tracev(__e, ...) tracev_event(TRACE_CLASS_EQ_IIR, __e, ##__VA_ARGS__)
+#define tracev_wi(comp_ptr, __e, ...)			\
 	tracev_event_comp(TRACE_CLASS_EQ_IIR, comp_ptr,		\
 			  __e, ##__VA_ARGS__)
 
-#define trace_eq_error(__e, ...) \
+#define trace_er(__e, ...) \
```
Comment (paulstelian97) [inline src/audio/eq_iir/eq_iir.c]: "er"? Too vague, is this ONLY to prevent lines too long? Then not OK in my view.

File: src/audio/eq_iir/eq_iir.c [hunk 3]
```diff
@@ -31,30 +31,31 @@
...
 	trace_error(TRACE_CLASS_EQ_IIR, __e, ##__VA_ARGS__)
-#define trace_eq_error_with_ids(comp_ptr, __e, ...)		\
+#define trace_ei(comp_ptr, __e, ...)		\
```
Comment (paulstelian97) [inline src/audio/eq_iir/eq_iir.c]: Same with "ei", I believe you should just leave it as-is.

File: src/audio/eq_iir/eq_iir.c [hunk 4]
```diff
@@ -339,38 +340,32 @@ static void eq_iir_free_delaylines(struct comp_data *cd)
...
+	      "%u responses", config->channels_in_config,
+	      config->number_of_responses);
 
 	/* Sanity checks */
 	if (nch > PLATFORM_MAX_CHANNELS ||
 	    config->channels_in_config > PLATFORM_MAX_CHANNELS ||
 	    !config->channels_in_config) {
-		trace_eq_error("eq_iir_setup() error: "
-			       "invalid nch or channels_in_config");
+		trace_er("eq_iir_init_coef(), error: invalid channels count");
```
Comment (paulstelian97) [inline src/audio/eq_iir/eq_iir.c]: Because this one with the original `trace_eq_error` shouldn't make Checkpatch complain, and if it does you may safely ignore it.

File: src/audio/eq_iir/eq_iir.c [hunk 5]
```diff
@@ -339,38 +340,32 @@ static void eq_iir_free_delaylines(struct comp_data *cd)
...
 		return -EINVAL;
 	}
 	if (config->number_of_responses > SOF_EQ_IIR_MAX_RESPONSES) {
-		trace_eq_error("eq_iir_setup() error: number_of_responses"
-			       " > SOF_EQ_IIR_MAX_RESPONSES");
+		trace_er("eq_iir_init_coef(), error: # of resp exceeds max");
```
Comment (paulstelian97) [inline src/audio/eq_iir/eq_iir.c]: Rewords of traces should be in a separate commit from the actual new functionality.

File: src/audio/eq_iir/eq_iir.c (hunk 1 above)
Comment (singalsu) [reply src/audio/eq_iir/eq_iir.c]: The obvious reason is avoiding long lines. Yes, I'll place them to separate commit.

File: src/audio/eq_iir/eq_iir.c (hunk 2 above)
Comment (singalsu) [reply src/audio/eq_iir/eq_iir.c]: Would tr_error() be better? Also when tracing an error could we omit string "error:" from them since we already have trace class for errors? All of these tricks help a lot to keep the lines <= 80 chars.

File: src/audio/eq_iir/eq_iir.c (hunk 3 above)
Comment (singalsu) [reply src/audio/eq_iir/eq_iir.c]: I think "_with_ids" is an ugly bloat. It should be the default trace type and use the original trace macro name. Another macro could be defined for functions without access to "dev" handle.

File: src/audio/eq_iir/eq_iir.c (hunk 1 above)
Comment (paulstelian97) [reply src/audio/eq_iir/eq_iir.c]: I honestly believe that this actually makes code less clear, so NAK on this one. Long lines are fine in certain circumstances (greppability). I'd rather sacrifice alignment than readability or allow lines of 81-82 characters on a case-by-case basis.

File: src/audio/eq_iir/eq_iir.c (hunk 2 above)
Comment (paulstelian97) [reply src/audio/eq_iir/eq_iir.c]: I'd say leave it as "trace_eq_error". You can omit the "error: " prefix if you wish. And then format strings are not subject to the <=80 chars restriction anyway.

File: src/audio/eq_iir/eq_iir.c (hunk 3 above)
Comment (paulstelian97) [reply src/audio/eq_iir/eq_iir.c]: I do agree that it should be the default but i'd take this into a different, more involved discussion rather than a "pass-by refactor". So you should instead make a separate PR where you would do such proposed changes to the traces so we can discuss them rather than blocking this PR with such issues.

File: src/audio/eq_iir/eq_iir.c (hunk 2 above)
Comment (singalsu) [reply src/audio/eq_iir/eq_iir.c]: OK, I'll revert and the long lines will be back. I hope the checkpatch.pl script can be in future updated to contain these rules relaxations. It would be nice to see again the column of all green check marks from CI.

Comment (paulstelian97) [reply src/audio/eq_iir/eq_iir.c]: Yeah, @jajanusz attempted to update certain things already (please rebase to latest master, I'm not sure you are using that update) so some "line too long" warnings shouldn't exist.

**Comment by singalsu** [comment]
> I won't block this one even though the fact that there are still changes in the trace messages bothers me. At least there isn't anything that becomes significantly less clear now.
> 
> While a cleanup of the changed messages from there would be best I can leave it as such.

I left into first commit trace prints changes those were due to function name change. The setup() function was split to init_coef() and init_delay() functions to be able to do those separately. The trace texts in init_coef() function those held the old function name were included to first commit.

Thanks for reviewing this @paulstelian97 ! Others please review too.

**Comment by singalsu** [comment]
File iir_generic.c was missing from first commit after a git reset. Hope it's now OK in this just pushed version.

**Comment by lgirdwood** [comment]
CI issues not relevant to this PR.

**Comment by juimonen** [comment]
@lgirdwood I think this PR is not recalculating coefficients, they are already calculated in user space way before... so I guess this is just switching them and clearing delay lines. @singalsu correct if I'm wrong.

**Comment by singalsu** [comment]
> @lgirdwood I think this PR is not recalculating coefficients, they are already calculated in user space way before... so I guess this is just switching them and clearing delay lines. @singalsu correct if I'm wrong.

Yep, all calculation is done in Octave EQ tool SOF/tools/tune/eq. However if need for realtime parameteric EQ the computation is not complicated, especially on some host side audio server for realtime updates. On FW it would be possible too but maybe with some simplifications into scaling and poles & zeros into biquads assign code.
